src/main.rs

```diff
@@ -66,47 +90,83 @@ fn main() -> Result<()> {
-                KeyCode::Char('k') => rotation(std::f32::consts::FRAC_PI_8 / 4.0, Vec3::j()),
-                KeyCode::Char('i') => rotation(-std::f32::consts::FRAC_PI_8 / 4.0, Vec3::j()),
-                KeyCode::Char('o') => rotation(std::f32::consts::FRAC_PI_8 / 4.0, Vec3::i()),
-                KeyCode::Char('u') => rotation(-std::f32::consts::FRAC_PI_8 / 4.0, Vec3::i()),
-                _ => (),
+            let mut did_rotation: bool = true;
+            if input.key_pressed(VirtualKeyCode::J) {
+                rotation(PI_FRAC_32, Vec3::k());
+            } else if input.key_pressed(VirtualKeyCode::L) {
+                rotation(-PI_FRAC_32, Vec3::k());
+            } else if input.key_pressed(VirtualKeyCode::K) {
+                rotation(PI_FRAC_32, Vec3::j());
+            } else if input.key_pressed(VirtualKeyCode::I) {
+                rotation(-PI_FRAC_32, Vec3::j());
+            } else if input.key_pressed(VirtualKeyCode::O) {
+                rotation(PI_FRAC_32, Vec3::i());
+            } else if input.key_pressed(VirtualKeyCode::O) {
+                rotation(-PI_FRAC_32, Vec3::i())
+            } else {
+                did_rotation = false;
             }
 
-            queue_render(&mut stdout, &world, &camera)?;
-            stdout.flush()?;
+            did_rotation || did_movement
+        } else {
+            false
+        };
+
+        // there is probably a way in the API to do this better but this was just very quick
+        let redraw_requested: bool = {
+            if let Event::RedrawRequested(_) = event {
+                true
+            } else {
+                false
+            }
+        };
+
+        // Draw the current frame
+        if keyboard_input || redraw_requested {
+            queue_render(pixels.get_frame_mut(), &world, &camera);
+            if pixels
+                .render()
+                .map_err(|e| panic!("pixels.render() failed: {}", e))
+                .is_err()
+            {
+                *control_flow = ControlFlow::Exit;
+                return;
+            }
         }
-    }
+    });
+}
 
-    execute!(stdout, terminal::LeaveAlternateScreen, cursor::Show)?;
+fn queue_render(frame: &mut [u8], world: &World, camera: &Camera) {
+    // Create a instant here to time how long it takes to render a frame
+    let now = std::time::Instant::now();
 
-    crossterm::terminal::disable_raw_mode()?;
+    // used to zip with frame data in place of enumerating (which cannot be done with par_chunks_exact_mut)
+    let index = 0..(DIMS.0 * DIMS.1);
 
-    Ok(())
-}
+    frame
+        .par_chunks_exact_mut(4)
+        .zip(index)
+        .for_each(|(pixel, i)| {
+            // (x,y) of pixel on screen
+            let (x, y): (i32, i32) = (((i % DIMS.0) as i32), ((i / DIMS.0) as i32));
+            let x_w = x as f32 - (DIMS.0 as f32) / 2.0;
+            let y_w = y as f32 - (DIMS.1 as f32) / 4.0;
```

Comment:
It wasn't clear but division by 4 here instead of 2 like the line above was because of the text rendering method.
```suggestion
            let y_w = y as f32 - (DIMS.1 as f32) / 2.0;
```